compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp

```diff
@@ -268,6 +268,170 @@ struct BubbleExpandThroughExtract final
   }
 };
 
+struct BubbleExpandThroughConcat final
+    : public OpRewritePattern<tensor::ExpandShapeOp> {
+  using OpRewritePattern<tensor::ExpandShapeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExpandShapeOp expandOp,
+                                PatternRewriter &rewriter) const override {
+    auto concatOp = expandOp.getSrc().getDefiningOp<tensor::ConcatOp>();
+    if (!concatOp) {
+      return failure();
+    }
+
+    // Get concat dimension and input/result types.
+    int64_t concatDim = concatOp.getDim();
+    SmallVector<ReassociationIndices, 4> reassoc =
+        expandOp.getReassociationIndices();
+    RankedTensorType expandedType = expandOp.getResultType();
+    ArrayRef<int64_t> expandedShape = expandedType.getShape();
+    ArrayRef<int64_t> srcShape = concatOp.getResultType().getShape();
+
+    // Find the "new" concat dim and if it is part of an expansion dim
+    int64_t srcRank = srcShape.size();
+    int64_t newConcatDim = -1, expandedDimCount = 0;
+    int64_t count = 0;
+    for (int64_t i = 0; i < srcRank; ++i) {
+      int64_t n = reassoc[i].size();
+      if (i == concatDim) {
+        newConcatDim = count;
+        expandedDimCount = n;
+        break;
+      }
+      count += n;
+    }
+    if (newConcatDim == -1)
+      return failure();
+
+    // Loop over all but the outermost expanded dims.
+    for (int64_t j = 1; j < expandedDimCount; ++j) {
+      int64_t expDim = expandedShape[newConcatDim + j];
+      if (ShapedType::isDynamic(expDim)) {
+        // If expanded dim is dynamic and not outermost, we cannot propagate.
+        return rewriter.notifyMatchFailure(
+            expandOp, "Expanded dim is dynamic and not outermost");
+      }
+
+      for (Value input : concatOp.getInputs()) {
+        // Check all static input shapes for divisibility.
+        auto inputType = dyn_cast<RankedTensorType>(input.getType());
+        if (!inputType) {
+          return rewriter.notifyMatchFailure(
+              expandOp, "Input to concat is not a RankedTensorType");
+        }
```

Comment:
```suggestion
        auto inputType = cast<RankedTensorType>(input.getType());
```

since `tensor.concat` requires RankedTensorType inputs